src/jdk.crypto.cryptoki/unix/native/libj2pkcs11/p11_md.c

```diff
@@ -262,20 +262,21 @@ JNIEXPORT jobject JNICALL Java_sun_security_pkcs11_wrapper_PKCS11_connect
 /*
  * Class:     sun_security_pkcs11_wrapper_PKCS11
  * Method:    disconnect
- * Signature: ()V
+ * Signature: (J)V
  */
-JNIEXPORT void JNICALL Java_sun_security_pkcs11_wrapper_PKCS11_disconnect
-    (JNIEnv *env, jobject obj)
-{
-    ModuleData *moduleData;
+JNIEXPORT void JNICALL Java_sun_security_pkcs11_wrapper_PKCS11_disconnect(
+        JNIEnv *env, jclass thisClass, jlong ckpNativeData) {
+
     TRACE0("DEBUG: disconnecting module...");
-    moduleData = removeModuleEntry(env, obj);
+    if (ckpNativeData != 0L) {
+        ModuleData *moduleData = jlong_to_ptr(ckpNativeData);
 
-    if (moduleData != NULL) {
-        dlclose(moduleData->hModule);
+        if (moduleData != NULL && moduleData->hModule != NULL) {
```

Comment:
The moduleData != NULL check seems to duplicate with the line 271? Otherwise looks fine.

Reply:
Hm, I see.  Updated.